man/man1/units.1
.TH UNITS 1NEMO  "20 November 2022"

.SH "NAME"
units \- units conversion utility

.SH "SYNOPIS"
\fBunits from=\fP\fIfrom\fP \fBto=\fP\fIto\fP  [parameter=value ...]

.SH "DESCRIPTION"
The units program converts quantities expression in various scales to 
their equivalents in other scales.  The units program can only
handle multiplicative (linear) scale changes.  This means
it cannot convert Centigrade to Fahrenheit, for example.

.SH "PARAMETERS"
The following parameters are recognized in any order if the keyword
is also given:
.TP 20
\fBfrom=\fP\fIunit_from\fP.
Unit to convert from. No default.
.TP
\fBto=\fP\fIunit_to\fP
Unit to convert to. No default.
.TP
\fBunit=\fP\fIunit_file\fP
Units database file. Multiple files can be given, separated by commas.
A search path "\fB.:$NEMODAT\fP" is employed, so a local
file is always tried before a file in \fB$NEMODAT\fP.
Default: \fBunits.lib\fP.
.TP
\fBshow=mul|div\fP
Option to select the multiplicative or it's inverse
factor you need to convert from \Iunit_from\fP to \fIunit_to\fP, i.e.
.nf
	unit_from = unit_to * mul
	unit_from = unit_to / div
.fi
Valid options are \fBmul\fP and \fBdiv\fP.
You can also select to show both, although
the \fBmul\fP is always shown first, irrespective of
the order given.
[Default: \fBmul\fP].
.TP
\fBquiet=t|f\fP
Be quiet in interactive mode? [Default: \fBt\fP]   
\fINOTE: interactive mode not implemented\fP

.SH "EXAMPLES"
It takes about 977 km/s to make one kpc in 10^6 yr:
.nf

    % units kpc/Myr km/s 
    977.81311

.fi
or a bit more mondaine:
.nf

    % units meters feet
    3.2808399

.fi
.PP
You can also use constants in the units expressions.
For example, if you need to know the length of something 
in multiples of \fI5 cm\fP, 
but know it's \fI3 feet\fP, try
.nf
    % units "3*feet" "5*cm"
      18.288 
.fi
Note the usage of the (single or double) quotes to prevent the shell
from interpreting the * meta character.
.PP
Powers of units can be specified using the '^' character as shown in
the example, or by simple concatenation: 'cm3' is equivalent to 'cm^3'.
Multiplication of units can be specified by using spaces, a dash or
an asterisk.  Division of units is indicated by the slash ('/').  
Note that multiplication has a higher precedence than division, so
'm/s/s' is the same as 'm/s^2' or 'm/s s'
If the user enters incompatible unit types, the units program will
print a message indicating that the units are not conformable and
it will display the reduced form for each unit:
.nf

    % units ergs/hour "fathoms kg^2 / day"
    conformability error
            2.7777778e-11 kg m^2 / sec^3
            2.1166667e-05 kg^2 m / sec
.fi
Also note the use of double quotes 
in order for NEMO to read the \fIunit_to\fP
in the second argument/keyword.
.LP
The conversion information is read from a units data file.  The default
file includes definitions for most familiar units, abbreviations and
metric prefixes.  Some constants of nature included are:
.in +4m
.ta
.ta 9m
.nf

pi	ratio of circumference to diameter
c	speed of light
e	charge on an electron
g	acceleration of gravity
force	same as g
mole	Avogadro's number
water	pressure per unit height of water
mercury	pressure per unit height of mercury
au	astronomical unit

.fi
.in -4m
\'Pound' is a unit of mass.  Compound names are run together
so 'poundforce' is a unit of force.  British units that differ from their
US counterparts are prefixed with 'br', and currency is prefixed with
its country name: 'belgiumfranc', 'britainpound'.  When searching for
a unit, if the specified string does not appear exactly as a unit
name, then the units program will try to remove a trailing 's' or
a trailing 'es' and check again for a match.  
.LP
All of these definitions can be read in the standard units file, or you
can supply your own file.  A unit is specified on a single line by
giving its name and an equivalence.  One should be careful to define
new units in terms of old ones so that a reduction leads to the
primitive units which are marked with '!' characters.  
The units program will not detect infinite loops that could be caused
by careless unit definitions.
.LP
Prefixes are defined in the same was as standard units, but with 
a trailing dash at the end of the prefix name.  
.PP
.nf
.ta +1i +1i

prefix	abbrev.	factor


quetta-	Q-	1e30
ronna-	R-	1e27
yotta-	Y-	1e24
zetta-	Z-	1e21
exa-	E-	1e18
peta-	P-	1e15
tera-	T-	1e12
giga-	G-	1e9
mega-	M-	1e6
myria-		1e4
kilo-	k-	1e3
hecto-	h-	1e2
deka-	da-	1e1
deci-	d-	1e-1
centi-	c-	1e-2
milli-	m-	1e-3
micro-		1e-6
nano-		1e-9
pico-	p-	1e-12
femto-	f-	1e-15
atto-	a-	1e-18
zopto-	z-	1e-21
yocto-	y-	1e-24
ronto-	r-	1e-27
quecto-	q-	1e-30

.fi
.SH G
The gravitational constant is often set to unity in N-body codes. However, for sake
of completeness, here are a few common values of G in some other 'common'
unit systems, recalling that G*m/(r*v*v) is dimensionless:
.nf
  mass   length   velocity    G            1/G            time

     g       cm       cm/s    6.6732e-8    1.49853e+07      s      (cgs)
    kg        m        m/s    6.6732e-11   1.49853e+10      s      (SI)
     M     Rsol       km/s    1.90809e5                            (solar tides)
     M       AU       km/s    886.8        1.128e-3        4.74yr  (planets)
     M       AU  29.8 km/s    1            1              0.159yr  (planets)
     M       AU               39.478                       1yr     (planets) 
     M       pc       km/s    4.302e-3     2.32385e2       1e6.yr  (starcluster)
     M       kpc      km/s                 2.32385e5       1e9.yr
1e10.M    10.kpc  100.km/s                 2.32385         1e8.yr  (galaxy)
~2e4.M       kpc ~990.km/s    1            1               1e6.yr  (galaxy)
1e11.M    10.kpc ~97.8.km/s   4.497        0.22237         1e8.yr  (galaxy)
1e10.M     1.kpc     1.km/s   43007.1                              (GADGET)
2.325e9 M  1.kpc   100.km/s   .            .               .       (galactICS?)
.fi

.SH "MEMORABLE CONVERSIONS"
.nf
.ta +2i
1 pc	3.086e18 cm
1 AU	1.496e13 cm
1 radian	206265 arcsec
1 yr       	3.156e7 sec
1 solarmass	1.988e33 g
1 km/s      	1.023 pc per million year
1 arcsec/yr	4.74 km/s at 1 pc
1 arcsec	4.848 pc/Mpc_distance

.fi
See also \fIconstants(5NEMO)\fP for some other constants.

.SH "BUGS"
.LP
The effect of including a '/' in a prefix is surprising.
.LP
Exponents entered by the user can be only one digit.
.LP
The user must use | to indicate division of numbers and / to
indicate division of symbols.  This distinction should not
be necessary.
.LP
The program contains various arbitrary limits on the length
of the units converted and on the length of the data file.

.SH "SEE ALSO"
expr(1), units(5NEMO), nemoinp(1NEMO)
.PP
NIST: http://physics.nist.gov/cuu/Constants/index.html 
.PP
NIST: http://physics.nist.gov/PhysRefData/contents.html
.PP
FRINK: http://futureboy.homeip.net/frinkdocs/
.PP
units calculator: http://www.astro.wisc.edu/~dolan/constants/calc.html
.PP
astropy units:  https://docs.astropy.org/en/stable/units/index.html
.PP
Granot, \fIScaling relations between numerical simulations and physical systems they represent\P, 
http://arxiv.org/pdf/1111.6865v1
.PP
https://en.wikipedia.org/wiki/N-body_units

.SH "FILES"
.nf
$NEMODAT/units.lib - the standard units library 
/usr/lib/units     - original units library from units(1)
.fi
Sadly, the modified units.c with the NEMO interface has been lost

.SH "AUTHOR"
Adrian Mariano (adrian@u.washington.edu or mariano@geom.umn.edu)

.SH "UPDATE HISTORY"
.nf
.ta +1.5i +5.5i
14-jul-93	V1.0 Original version	AM
21-jul-93	V2.0 Conversion to nemo	PJT
8-aug-93	V2.1 cleanup, reset defaults, switch key order	PJT
20-nov-2022	V2.2 added r/q/R/Q as per new NIST decree	PJT
.fi


src/nbody/evolve/aarseth/nbody4/units.f
      SUBROUTINE UNITS
*
*
*       Initialization of units & scaling factors.
*       ------------------------------------------
*
      INCLUDE 'common4.h'
*
      
*       Define GM & PC in cgs units and AU in pc.
      GM = 6.67D-08*1.989D+33
      PC = 3.0857D+18
      AU = 2.0627E+05
*
*       Form scaling factors for binary periods A*SQRT(A/M) to yrs and days.
      YRS = (RBAR*AU)**1.5/SQRT(ZMBAR)
      DAYS = 365.24*YRS
*
*       Specify conversion factors for lengths to solar radii & AU.
      SU = PC/(AU*6.96D+10)*RBAR*AU
      RAU = RBAR*AU
*
*       Copy solar mass scaling to new variable (M = BODY*<M>).
      SMU = ZMBAR
*
*       Form time scale in seconds and velocity scale in km/sec.
      TSTAR = SQRT(PC/GM)*PC
      VSTAR = 1.0D-05*SQRT(GM/PC)
*
*       Convert time scale from units of seconds to million years.
      TSTAR = TSTAR/(3.15D+07*1.0D+06)
*
*       Ensure ZMBAR & RBAR > 0 (=0: assume <M>/Sun = 1, RBAR = 1 pc).
      IF (ZMBAR.LE.0.0D0) ZMBAR = FLOAT(N)/ZMASS
      IF (RBAR.LE.0.0D0) RBAR = 1.0
*
*       Scale to working units of RBAR in pc & ZMBAR in solar masses.
      TSTAR = TSTAR*SQRT(RBAR**3/(ZMASS*ZMBAR))
      VSTAR = VSTAR*SQRT(ZMASS*ZMBAR/RBAR)
*
*       Copy TSTAR to secondary time-scale factor.
*     TSTAR = 2.62d0*TSTAR
      TSCALE = TSTAR
*
*       Physical scaling: X, M, V, T from RBAR*X, ZMBAR*M, VSTAR*V, TSTAR*T.
      WRITE (6,10)  RBAR, ZMBAR, VSTAR, TSTAR, BODYM*ZMBAR, SU
   10 FORMAT (/,12X,'PHYSICAL SCALING:    R* =',F4.1,'  M* =',F8.1,
     &              '  V* =',F6.3,'  T* =',F6.3,'  <M> =',F5.2,
     &              '  SU =',1P,E8.1,/)
*
      RETURN
*
      END
